### PR TITLE
WIP archive_load_files

### PR DESCRIPTION
**Pull request by llehtinen**
## Problem

https://transferwise.atlassian.net/browse/AP-1011

## Proposed changes

https://docs.google.com/document/d/11UTlmWVJS9aGickmyxpXOUhrv4RTPnm8zip_IqeR2J0/edit?ts=609e4106

## Types of changes

What types of changes does your code introduce to PipelineWise?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] Description above provides context of the change
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Unit tests for changes (not needed for documentation changes)
- [ ] CI checks pass with my changes
- [ ] Bumping version in `setup.py` is an individual PR and not mixed with feature or bugfix PRs
- [ ] Commit message/PR title starts with `[AP-NNNN]` (if applicable. AP-NNNN = JIRA ID)
- [ ] Branch name starts with `AP-NNN` (if applicable. AP-NNN = JIRA ID)
- [ ] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [ ] Relevant documentation is updated including usage instructions

### CODE REVIEW

File: target_snowflake/upload_clients/base_upload_client.py [hunk 1]
```diff
@@ -24,3 +24,9 @@ def delete_object(self, stream: str, key: str) -> None:
         """
         Delete object
         """
+
+    @abstractmethod
+    def copy_object(self, source_key: str, target_key: str, target_metadata: dict) -> None:
```
Comment (koszti) [inline target_snowflake/upload_clients/base_upload_client.py:29]: The [SnowflakeUploadClient](https://github.com/transferwise/pipelinewise-target-snowflake/blob/master/target_snowflake/upload_clients/snowflake_upload_client.py) doesn't implement this.

The `copy_object` acceptable only for s3_upload_client but I think an implementation is still required in the SnowflakeUploadClient as well. Maybe we can raise `NotImplementedError` with some human readable error message.

File: target_snowflake/__init__.py [hunk 2]
```diff
@@ -170,6 +171,31 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
             else:
                 records_to_load[stream][primary_key_string] = o['record']
 
+            if config.get('archive_load_files.enabled'):
+                # keep track of min and max
+                archive_load_files_primary_column =\
+                    config.get('archive_load_files.primary_column') or primary_key_string
```
Comment (koszti) [inline target_snowflake/__init__.py:177]: this should be done once when processing the `SCHEMA` message and not at every `RECORD` message.

Primary key column(s) is available in the schema message as `o['key_properties']` and the incremental key(s) as `o['bookmark_properties'].

File: target_snowflake/__init__.py [hunk 3]
```diff
@@ -170,6 +171,31 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
...
+                archive_primary_column_value =\
+                    records_to_load[stream][primary_key_string][archive_load_files_primary_column]
+                values = archive_load_files_data[stream] or {
+                    's3_key': 'tbd',  # TODO: Turn archive_load_files.naming_convention into actual key
+                    'database': 'tbd',  # TODO: Where to get this
+                    'schema': 'tbd',  # TODO: Where to get this
+                    'table': stream,
```
Comment (koszti) [inline target_snowflake/__init__.py:184]: these values, especiall`s3_key` should be defined at an another place level when loading into s3.

The stream is a concatenated string of db-schema and db-table and that's probably enough  to pass furhter to other functions. The `database` is  not acceptable in general in target-snowflake. One ppw stream can read only from one database and DB-schema-table hierarchy is not acceptable to every db type.  (pg vs mariadb)

File: target_snowflake/__init__.py [hunk 4]
```diff
@@ -170,6 +171,31 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
             else:
                 records_to_load[stream][primary_key_string] = o['record']
 
+            if config.get('archive_load_files.enabled'):
```
Comment (koszti) [inline target_snowflake/__init__.py:174]: we also need to check if we're using the `S3UploadClient` and not the `SnowflakeUploadClient`. Archiving load files is acceptable only if we use the `S3UploadClient`.

The `SnowflakeUploadClient` is using snowflake managed table stages and we can't archive anything onto it.

There is a check at around https://github.com/transferwise/pipelinewise-target-snowflake/blob/ad0d71baa9dee4c47b05869b7ca6c333593f8da3/target_snowflake/db_sync.py#L273

File: target_snowflake/__init__.py (hunk 3 above)
Comment (llehtinen) [reply target_snowflake/__init__.py:184]: @koszti One issue with `s3_key` is that it requires access to the config. From here, currently, `config` is not passed further than `flush_streams`. Would you prefer to have the entire config propagated further from there, or add the `archive_load_files` config as init param to the `S3UploadClient`, or something else?

File: target_snowflake/__init__.py [hunk 5]
```diff
@@ -257,6 +276,26 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
                 else:
                     stream_to_sync[stream] = DbSync(config, o, table_cache, file_format_type)
 
+                if archive_load_files_enabled:
+                    # Determine archive_load_files_primary_column
+                    # 1) Use incremental replication key if defined
+                    # 2) Otherwise use primary key
+                    if False and 'bookmark_properties' in o and len(o['bookmark_properties']) > 0:
```
Comment (llehtinen) [inline target_snowflake/__init__.py:283]: TBD: What's the proper way to identify incremental replication key

File: target_snowflake/__init__.py [hunk 6]
```diff
@@ -113,6 +113,9 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
     batch_size_rows = config.get('batch_size_rows', DEFAULT_BATCH_SIZE_ROWS)
     batch_wait_limit_seconds = config.get('batch_wait_limit_seconds', None)
     flush_timestamp = datetime.utcnow()
+    archive_load_files_enabled =\
```
Comment (llehtinen) [inline target_snowflake/__init__.py:116]: Think of better way to check the config - if archive_load_files.enabled but s3 bucket isn't defined, raise exception

Comment (koszti) [reply target_snowflake/__init__.py:116]: Config validation is at https://github.com/transferwise/pipelinewise-target-snowflake/blob/master/target_snowflake/db_sync.py#L65 , where we can do something like this:

```
    # Check if archive load files option is using external stages
    archive_load_files = config.get('archive_load_files', {})
    if archive_load_files.get('enabled') and not config.get('s3_bucket', None):
        errors.append('Archive load files option can be used only with external s3 stages. Please define s3_bucket')
```

Please note that separating config variables with dots are not working: `db_conn.s3_bucket` and `archive_load_files.enabled`. We need to access them as nested dictionaries.

File: target_snowflake/__init__.py (hunk 5 above)
Comment (koszti) [reply target_snowflake/__init__.py:283]: check [messages-simple-table.json](https://github.com/transferwise/pipelinewise-target-snowflake/blob/ad0d71baa9dee4c47b05869b7ca6c333593f8da3/tests/integration/resources/messages-simple-table.json), maybe this is a better example.

The incremental key is defined in details only in the `STATE` message, but taps are not sending them necessarily always after the `SCHEMA` and before the `RECORD` messages. The order is not defined by the [singer-spec](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#state-message)

But taps are copying replication key into the `bookmark_properties`. If it's logical replication then tap-postgres using the `lsn` keyword. 

I think the best what we can do is to get `o['bookmark_properties']` and if the result is one of the column in the schema then we use it as `archive_load_files_primary_column`. If not exists then we use the PK: o['key_properties'][0].

Incremental key is a single value, but primary key can have multiple columns

File: target_snowflake/__init__.py [hunk 7]
```diff
@@ -257,6 +276,26 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
...
+                        LOGGER.info("Using %s from bookmark_properties as archive_load_files_primary_column",
+                                    o['bookmark_properties'][0])
+                        archive_load_files_primary_column = o['bookmark_properties'][0]  # TBD Can there be more than 1?
+                    else:
+                        LOGGER.info("Using %s from key_properties as archive_load_files_primary_column",
+                                    o['key_properties'][0])
+                        archive_load_files_primary_column = o['key_properties'][0]  # Behavior when key multi-col?
+
+                    archive_load_files_data[stream] = {
+                        'tap': config.get('id'),
```
Comment (koszti) [inline target_snowflake/__init__.py:293]:
```suggestion
                        'tap': config.get('tap_id'),
```

target-snowflake can't access the original YAML file. The config is a generated JSON of the items under `db_conn` and some other propertes listed [here](https://github.com/transferwise/pipelinewise/blob/eca52d4fa566f27c53b206a59ada86ee7d986a68/pipelinewise/cli/config.py#L320)

File: target_snowflake/__init__.py [hunk 8]
```diff
@@ -170,6 +173,20 @@ def persist_lines(config, lines, table_cache=None, file_format_type: FileFormatT
             else:
                 records_to_load[stream][primary_key_string] = o['record']
 
+            if archive_load_files_enabled and stream in archive_load_files_data:
+                # Keep track of min and max of the designated column
+                values = archive_load_files_data[stream]
+                archive_primary_column_name = values['column']
+                archive_primary_column_value = o['record'][archive_primary_column_name]
```
Comment (koszti) [inline target_snowflake/__init__.py:180]: What we should return if `archive_primary_column_name` is a composite primary key? How we should track it?

File: tests/unit/test_target_snowflake.py [hunk 9]
```diff
@@ -85,4 +85,86 @@ def test_persist_40_records_with_batch_wait_limit(self, dbSync_mock, flush_strea
...
-        assert flush_streams_mock.call_count == 41
\ No newline at end of file
+        assert flush_streams_mock.call_count == 41
+
+    @patch('target_snowflake.DbSync')
+    @patch('target_snowflake.flush_streams')
+    def test_archive_load_files(self, flush_streams_mock, dbSync_mock):
+        self.config['id'] = 'test-tap-id'
+        self.config['archive_load_files.enabled'] = True
+        self.config['db_conn.s3_bucket'] = 'dummy_bucket'
```
Comment (koszti) [inline tests/unit/test_target_snowflake.py:95]:
```suggestion
        self.config['s3_bucket'] = 'dummy_bucket'
```

Separating by dots is not valid and here we don't use the original YAML file. The config is a generated JSON of the items under `db_conn` and some other propertes listed [here](https://github.com/transferwise/pipelinewise/blob/eca52d4fa566f27c53b206a59ada86ee7d986a68/pipelinewise/cli/config.py#L320)

File: tests/unit/test_target_snowflake.py [hunk 10]
```diff
@@ -85,4 +85,86 @@ def test_persist_40_records_with_batch_wait_limit(self, dbSync_mock, flush_strea
...
+                'max': 20
+            }
+        }
+
+    @patch('target_snowflake.DbSync')
+    @patch('target_snowflake.load_stream_batch')
+    def test_archive_load_files_2(self, load_stream_batch_mock, dbSync_mock):
+        self.config['id'] = 'test-tap-id'
+        self.config['archive_load_files.enabled'] = True
+        self.config['db_conn.s3_bucket'] = 'dummy_bucket'
```
Comment (koszti) [inline tests/unit/test_target_snowflake.py:123]: same as above

File: tests/unit/test_target_snowflake.py [hunk 11]
```diff
@@ -85,4 +85,86 @@ def test_persist_40_records_with_batch_wait_limit(self, dbSync_mock, flush_strea
...
+            'min': 1,
+            'max': 20
+        }
+
+    @patch('target_snowflake.DbSync')
+    @patch('target_snowflake.os.remove')
+    def test_archive_load_files_3(self, os_remove_mock, dbSync_mock):
+        self.config['id'] = 'test-tap-id'
+        self.config['archive_load_files.enabled'] = True
+        self.config['db_conn.s3_bucket'] = 'dummy_bucket'
```
Comment (koszti) [inline tests/unit/test_target_snowflake.py:147]: same as above

File: target_snowflake/db_sync.py [hunk 12]
```diff
@@ -63,6 +63,11 @@ def validate_config(config):
     if not config_default_target_schema and not config_schema_mapping:
         errors.append("Neither 'default_target_schema' (string) nor 'schema_mapping' (object) keys set in config.")
 
+    # Check if archive load files option is using external stages
+    archive_load_files = config.get('archive_load_files', {})
+    if archive_load_files.get('enabled') and not config.get('s3_bucket', None):
```
Comment (llehtinen) [inline target_snowflake/db_sync.py:68]: More specific assertion on (see how other booleans, if any, are evaluated)
```
archive_load_files.get('enabled')
```

File: target_snowflake/stream_utils.py [hunk 13]
```diff
@@ -114,3 +114,16 @@ def stream_name_to_dict(stream_name, separator='-'):
         'schema_name': schema_name,
         'table_name': table_name
     }
+
+
+def get_archive_load_files_primary_column(singer_msg: Dict):
+    """Derive archive load files primary column from a Singer message dictionary"""
+    if singer_msg['type'] != "SCHEMA":
+        raise Exception("Expecting type SCHEMA, got {}".format(singer_msg['type']))
```
Comment (llehtinen) [inline target_snowflake/stream_utils.py:122]: Use custom exception

**Comment by koszti** [comment]
this PR is obsolete and covered by https://github.com/transferwise/pipelinewise-target-snowflake/pull/178

**Pull request closed without merging**